sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1185,7 +1186,7 @@ object ReplaceDistinctWithAggregate extends Rule[LogicalPlan] {
  */
 object ReplaceDeduplicateWithAggregate extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = plan transform {
-    case Deduplicate(keys, child, streaming) if !streaming =>
+    case d @ Deduplicate(keys, child, _) if d.originalOutputMode != OutputMode.Append() =>
```

Comment:
existing: It's kind of odd that this decision is made in the optimizer and not the query planner.  I think our aggregate operator is actually worse than the specialized deduplication operator (since the specialized one is non-blocking).

It doesn't have to be in this PR, but we should probably move this to the planner eventually.